crates/bevy_ecs/src/lib.rs

```diff
@@ -1,5 +1,7 @@
 // FIXME(11590): remove this once the lint is fixed
 #![allow(unsafe_op_in_unsafe_fn)]
+// TODO: remove once Edition 2024 is released
+#![allow(dependency_on_unit_never_type_fallback)]
```

Comment:
I actually prefer having the explicit allow with a `TODO`. Once 2024 releases, we will want to go back over this and potentially implement the relevant traits for systems that output `!`. If we don't, users who try to add similar `todo!` closures to their apps will fail to compile.

Reply:
I agree, and even if I was overruled, I'd rather not make this change as part of this PR. It's really unrelated and due only to an annoying (and imo mistaken) choice from the rust lang team.

Reply:
> Once 2024 releases, we will want to go back over this and potentially implement the relevant traits for systems that output `!`.

You could implement those traits now, right?  The 2024 edition will make those impls more useful, but you can create them before then so that you don't have to scramble after the release.  (Not suggesting that we should do that in this PR!  Just noting that we don't have to wait for the edition.)  

> I'd rather not make this change as part of this PR

Sorry, I wasn't trying to make more work for you!  Your message on Discord made it seem like you wanted to resolve the warning but couldn't, and I was just trying to offer a way to do so.  :)

Reply:
> You could implement those traits now, right?

Sadly not! The `!` type isn't stabilized until Edition 2024, so we can't make the "proper" fix until then.

Reply:
Huh, I didn't see that in the edition guide. ... Oh, I see, it's even worse than that: the plan is to stabilize it sometime *after* the edition.  

... wait, then what are we going to do when the new edition is out?  The code with the warnings won't compile on Rust 2024 without impls for `!`, but we can't add those impls until `!` is stabilized at some later date.

Reply:
Exactly. This is why I am a tad frustrated with the rust team. Theoretically this should not be an issue until bevy itself decides to upgrade, since each crate gets to decide its edition individually.

Reply:
I don't think I follow.  What do you think the rust team should have done differently?  This looks like a standard case of code that will fail to compile on the next edition causing a warning on the current one.  It failed CI because that builds with `-Dwarnings`, but if you ignore the warning then it still builds with the old semantics.

Reply:
The big issue as I see it is there'll be a period of time where the only way to accept a function that returns `!` as a parameter is to accept _all_ return types. So in this PR, we have no way of actually accepting a closure like `|| todo!()` as a system, since it returns an unnameable type in Edition 2024. Combined with the statement that they intend on back-porting this change to earlier editions makes things very complicated.

_If_ Edition 2024 launches with the never-type _and_ this change to inference, then it's fine IMO. But since it'll only launch with the inference change, we'll have this transitional period where closures are just worse.